// genezio/build/src/packageManagers/packageManager.js
import NpmPackageManager from "./npm.js";
import PnpmPackageManager from "./pnpm.js";
import YarnPackageManager from "./yarn.js";
export var PackageManagerType;
(function (PackageManagerType) {
    PackageManagerType["npm"] = "npm";
    PackageManagerType["yarn"] = "yarn";
    PackageManagerType["pnpm"] = "pnpm";
})(PackageManagerType || (PackageManagerType = {}));
/*
 * A mapping between package manager types and package manager instances.
 *
 * Note: This code is also used to warn if a package manager has been declared, but does not have
 * an implementation. (using the type checker)
 */
export const packageManagers = {
    npm: new NpmPackageManager(),
    yarn: new YarnPackageManager(),
    pnpm: new PnpmPackageManager(),
};
/*
 * This is the currenly selected package manager.
 */
let packageManager = packageManagers.npm;
export function setPackageManager(packageManagerType) {
    packageManager = packageManagers[packageManagerType] ?? packageManagers.npm;
}
export default packageManager;
