// genezio/build/src/utils/decorators/baseDecoratorExtractor.js
export class DecoratorExtractor {
    createGenezioMethodInfo(methodName, commentText) {
        const genezioMethodArguments = commentText
            .split("genezio:")[1]
            .split(" ")
            .filter((arg) => arg !== "");
        if (genezioMethodArguments.length < 1) {
            return;
        }
        let decoratorArguments = {};
        switch (genezioMethodArguments[0]) {
            case "http": {
                decoratorArguments = {
                    type: "http",
                };
                break;
            }
            case "cron": {
                decoratorArguments = {
                    type: "cron",
                    cronString: genezioMethodArguments.slice(1).join(" "),
                };
                break;
            }
            default: {
                decoratorArguments = {
                    type: "jsonrpc",
                };
                break;
            }
        }
        const methodInfo = {
            name: methodName,
            decorators: [
                {
                    name: "GenezioMethod",
                    arguments: decoratorArguments,
                },
            ],
        };
        return methodInfo;
    }
    createGenezioClassInfo(className, file, commentText) {
        const genezioDeployArguments = commentText
            .split("genezio: deploy")[1]
            .split(" ")
            .filter((arg) => arg !== "");
        let classType = "jsonrpc";
        switch (genezioDeployArguments[0]) {
            case "http": {
                classType = "http";
                break;
            }
            case "cron": {
                classType = "cron";
                break;
            }
        }
        const classInfo = {
            path: file,
            name: className,
            decorators: [
                {
                    name: "GenezioDeploy",
                    arguments: {
                        type: classType,
                    },
                },
            ],
            methods: [],
        };
        return classInfo;
    }
}
